src/stages/stage3_events.c

```diff
@@ -741,939 +81,689 @@ Boss* stage3_spawn_scuttle(cmplx pos) {
-	if(time < 0)
-		return;
-
-	if(render) {
-		r_draw_sprite(&(SpriteParams) {
-			.sprite = "fairy_circle",
-			.rotation.angle = DEG2RAD * 7 * time,
-			.scale.both = 0.7,
-			.color = RGBA(0.8, 1.0, 0.4, 0),
-			.pos = { creal(e->pos), cimag(e->pos) },
-		});
-	} else if(time % 5 == 0) {
-		tsrand_fill(2);
-		PARTICLE(
-			.sprite = "smoothdot",
-			.pos = 5*cexp(2*I*M_PI*afrand(0)),
-			.color = RGBA(0.6, 0.6, 0.5, 0),
-			.draw_rule = Shrink,
-			.rule = enemy_flare,
-			.timeout = 60,
-			.args = {
-				0.3*cexp(2*M_PI*I*afrand(1)),
-				add_ref(e),
-			},
-		);
-	}
+static void stage3_dialog_post_boss(void) {
+	PlayerMode *pm = global.plr.mode;
+	INVOKE_TASK_INDIRECT(Stage3PostBossDialog, pm->dialog->Stage3PostBoss);
 }
 
-static int wriggle_rocket_laserbullet(Projectile *p, int time) {
-	if(time == EVENT_DEATH) {
-		free_ref(p->args[0]);
-		return ACTION_ACK;
-	} else if(time < 0) {
-		return ACTION_ACK;
-	}
 
-	if(time >= creal(p->args[1])) {
-		if(p->args[2]) {
-			cmplx dist = global.plr.pos - p->pos;
-			cmplx accel = (0.1 + 0.2 * (global.diff / (float)D_Lunatic)) * dist / cabs(dist);
-			float deathtime = sqrt(2*cabs(dist)/cabs(accel));
-
-			Laser *l = create_lasercurve2c(p->pos, deathtime, deathtime, RGBA(0.4, 0.9, 1.0, 0.0), las_accel, 0, accel);
-			l->width = 15;
-
-			PROJECTILE(
-				.proto = p->proto,
-				.pos = p->pos,
-				.color = &p->color,
-				.draw_rule = p->draw_rule,
-				.rule = wriggle_rocket_laserbullet,
-				.args = {
-					add_ref(l),
-					deathtime,
-				}
-			);
+/*
+ * Stage mobs
+ */
 
-			play_sound("redirect");
-			play_sound("shot_special1");
-		} else {
-			int cnt = 22, i;
-			float rot = (global.frames - global.boss->current->starttime) * 0.0037 * (global.diff);
-			Color *c = HSLA(fmod(rot, M_PI*2)/(M_PI/2), 1.0, 0.5, 0);
+// burst swirls
 
-			for(i = 0; i < cnt; ++i) {
-				float f = (float)i/cnt;
+TASK(death_burst, { BoxedEnemy e; int shot_type; }) {
+	// explode in a hail of bullets
+	Enemy *e = TASK_BIND(ARGS.e);
 
-				PROJECTILE(
-					.proto = pp_thickrice,
-					.pos = p->pos,
-					.color = c,
-					.rule = asymptotic,
-					.args = {
-						(1.0 + psin(M_PI*18*f)) * cexp(I*(2.0*M_PI*f+rot)),
-						2 + 2 * global.diff
-					},
-				);
-			}
+	float r, g;
+	if(frand() > 0.5) {
+		r = 0.3;
+		g = 1.0;
+	} else {
+		r = 1.0;
+		g = 0.3;
+	}
 
-			PARTICLE(
-				.proto = pp_blast,
-				.pos = p->pos,
-				.color = c,
-				.timeout = 35 - 5 * frand(),
-				.draw_rule = GrowFade,
-				.args = { 0, 1 + 0.5 * frand() },
-				.angle = M_PI * 2 * frand(),
-			);
-
-			// FIXME: better sound
-			play_sound("enemydeath");
-			play_sound("shot1");
-			play_sound("shot3");
-		}
+	int cnt = difficulty_value(24, 20, 16, 12);
+	for(int i = 0; i < cnt; ++i) {
+		double a = (M_PI * 2.0 * i) / cnt;
+		cmplx dir = cdir(a);
+		log_debug("shot_type: %d", ARGS.shot_type);
 
-		return ACTION_DESTROY;
+		PROJECTILE(
+				.proto = ARGS.shot_type ? pp_ball : pp_rice,
```

Comment:
If this is the only thing that depends on `shot_type`, you can pass the `ProjPrototype*` as a task argument directly:

```c
TASK(death_burst, { BoxedEnemy e; int ProjPrototype *shot_proj; }) {
	// ...
	PROJECTILE(
		.proto = ARGS.shot_proj,
	// ...
}
```